zmq/auth/thread.py

```diff
@@ -6,101 +6,92 @@
 # Copyright (C) PyZMQ Developers
 # Distributed under the terms of the Modified BSD License.
 
+import asyncio
 import logging
+import threading
 from itertools import chain
 from threading import Event, Thread
 from typing import Any, Dict, List, Optional, TypeVar, cast
 
 import zmq
+import zmq.asyncio
 from zmq import NOBLOCK, POLLIN
 from zmq.utils import jsonapi
 
+from .asyncio import AsyncioAuthenticator
 from .base import Authenticator
 
 
 class AuthenticationThread(Thread):
     """A Thread for running a zmq Authenticator
 
-    This is run in the background by ThreadedAuthenticator
+    This is run in the background by ThreadAuthenticator
     """
 
+    pipe: zmq.asyncio.Socket
+    loop: asyncio.AbstractEventLoop
+
+    poller: Optional[zmq.asyncio.Poller]
+
     def __init__(
         self,
-        context: "zmq.Context",
         endpoint: str,
         encoding: str = 'utf-8',
         log: Any = None,
-        authenticator: Optional[Authenticator] = None,
     ) -> None:
         super().__init__()
-        self.context = context or zmq.Context.instance()
```

Comment:
context ~always has to be passed here, because the transport is `inproc://`, the (underlying libzmq) context must be the same. I think that's why you're getting hangs (send on a socket with inproc will hang until the peer arrives).

I need to think of a better way to do this, but you can make an async reference to an existing context with:

```python
actx = zmq.asyncio.Context.shadow(ctx.underlying)
async_socket = actx.socket(zmq.PULL)
```

Next, you'll really want to do everything async within a coroutine, so `run` should look like:

```python
    def run(self) -> None:
        """Start the Authentication Agent thread task"""
        loop = asyncio.new_event_loop()
        try:
            loop.run_until_complete(self._run())
        finally:
            loop.close()
            # cleanup sockets

    async def _run(self) -> None:
        context = zmq.asyncio.Context.shadow(self.context.underlying)
        loop = asyncio.get_event_loop()
        ...

```